### PR TITLE
Azure Terraform module AKS cluster

### PR DESCRIPTION
**Pull request by chaitu6022**
What this PR does?

Terraform module for Azure AKS, this module will create a Virtual Network (Vnet), Subnets, Network Security Groups, Managed Identities, Storage Account, Azure Kubernetes Service cluster, AKS default node pool.

Issue Number

#7

┆Issue is synchronized with this [Jira Bug](https://k8ssandra.atlassian.net/browse/K8SSAND-531) by [Unito](https://www.unito.io)

### CODE REVIEW

File: azure/README.md [hunk 1]
```diff
@@ -1,3 +1,178 @@
...
+## Prerequisites
+
+|       NAME          |   Version  | 
+|---------------------|------------|
+| Terraform version   |   0.14     |
+| Azurerm provider    |   ~>2.49.0 |
+| Helm version        |   v3.5.3   |
+|   AZ CLI            |  ~>2.22.1  |   
+|  kubectl            |  ~>1.17.17 |
+|  python             |    3       |
```
Comment (bradfordcp) [inline azure/README.md]: Is there a reason these items have extra indentation?

```suggestion
| AZ CLI              |  ~>2.22.1  |   
| kubectl             |  ~>1.17.17 |
| python              |    3       |
```

File: azure/README.md [hunk 2]
```diff
@@ -1,3 +1,178 @@
...
+export TF_VAR_environment=<ENVIRONMENT_REPLACEME>
+ex:- export TF_VAR_environment=dev
+
+# Resource name prefix
+export TF_VAR_name=<CLUSTERNAME_REPLACEME>
+ex:- export TF_VAR_name=k8ssandra
+
+# Location
+export TF_VAR_region=<REGION_REPLACEME>
+ex:- export TF_VAR_region=us-east-1
```
Comment (bradfordcp) [inline azure/README.md]: Replace us-east-1 with an actual Azure region

```suggestion
ex:- export TF_VAR_region=eastus
```

File: azure/modules/aks/variables.tf [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+# Variables for the default node_pool.
+variable "private_subnet" {
+  description = "The subnet id of the virtual network where the virtual machines will reside."
+  type        = string
+}
+
+variable "vm_size" {
+  description = "Specifies the size of the virtual machine."
+  default     = "Standard_DS2_v2"
```
Comment (bradfordcp) [inline azure/modules/aks/variables.tf]: This instance type is too small.

```suggestion
  default     = "Standard_E8_v4"
```

File: azure/modules/aks/variables.tf [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+variable "min_count" {
+  description = "Minimum Node Count"
+  default     = 3
+  type        = number
+}
+variable "max_count" {
+  description = "Maximum Node Count"
+  default     = 5
```
Comment (bradfordcp) [inline azure/modules/aks/variables.tf]: Is this 5 total or per AZ?

File: azure/modules/iam/main.tf [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  ]
+}
+
+resource "azurerm_role_assignment" "role_assignment" {
+  name               = "00000000-0000-0000-0000-000000000000"
+  scope              = "Storage" # storage.id
+  role_definition_id = azurerm_role_definition.role_definition.role_definition_resource_id
+  principal_id       = azurerm_user_assigned_identity.user_assigned_identity.principal_id
+}
+*/
```
Comment (bradfordcp) [inline azure/modules/iam/main.tf]: Why is all of this commented out code present?

File: azure/modules/storage/README.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
+# Terraform google cloud storage module 
+This is a Dynamic modules in Terraform to create Azure Storage Account in a subnet.
+
+* main.tf : contains all the resources which will be created with `terraform apply` command. 
+* variables.tf : contains all variables required to create the resources.
+* outputs.tf : contains output attributes of the resources. 
+
+## google cloud resources created
```
Comment (bradfordcp) [inline azure/modules/storage/README.md]: This is an Azure module
```suggestion
## Azure resources created
```

File: azure/modules/storage/README.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
+# Terraform google cloud storage module
```
Comment (bradfordcp) [inline azure/modules/storage/README.md]: This is an Azure module

```suggestion
# Terraform Azure Blob Storage module 
```

File: azure/modules/storage/README.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+## Inputs
+
+| Name | Description | Type | Default | Required |
+|------|-------------|------|---------|:--------:|
+| <a name="input_account_tier"></a> [account\_tier](#input\_account\_tier) | The Storage Acount tier. | `string` | `"standard"` | no |
+| <a name="input_environment"></a> [environment](#input\_environment) | Name of the environment where infrastructure being built. | `string` | n/a | yes |
+| <a name="input_location"></a> [location](#input\_location) | Azure location where all the resources being created. | `string` | n/a | yes |
+| <a name="input_name"></a> [name](#input\_name) | Name is the prefix to use for resources that needs to be created. | `string` | n/a | yes |
+| <a name="input_private_subnet"></a> [private\_subnet](#input\_private\_subnet) | The subnet id of the virtual network where the virtual machines will reside. | `string` | n/a | yes |
+| <a name="input_replication_type"></a> [replication\_type](#input\_replication\_type) | The Storage Account Repluucation type. | `string` | `"LRS"` | no |
```
Comment (bradfordcp) [inline azure/modules/storage/README.md]:
```suggestion
| <a name="input_replication_type"></a> [replication\_type](#input\_replication\_type) | The Storage Account Replication type. | `string` | `"LRS"` | no |
```

File: gcp/env/variables.tf [hunk 9]
```diff
@@ -18,7 +18,7 @@ variable "name" {
 }
 
 variable "environment" {
-  description = "The environment of the infrastructure being built."
+  description = "Name of the environment where infrastructure being built."
```
Comment (bradfordcp) [inline gcp/env/variables.tf]: Should this happen in a separate PR?

File: gcp/modules/gcs/variables.tf [hunk 10]
```diff
@@ -18,7 +18,7 @@ variable "name" {
 }
 
 variable "environment" {
-  description = "The environment of the infrastructure being built."
+  description = "Name of the environment where infrastructure being built."
```
Comment (bradfordcp) [inline gcp/modules/gcs/variables.tf]: Should this happen in a separate PR?

**Review by bradfordcp — CHANGES_REQUESTED** [review]

File: azure/modules/iam/main.tf (hunk 5 above)
Comment (chaitu6022) [reply azure/modules/iam/main.tf]: I kept there for testing, removed from the repo. I will add another PR, if any of those blocks of commented code is required to add for Medusa.

File: azure/modules/aks/variables.tf (hunk 4 above)
Comment (chaitu6022) [reply azure/modules/aks/variables.tf]: The value of the node_count should be between the range of max_count and min_count in the container node pool.

File: azure/modules/iam/main.tf (hunk 5 above)
Comment (bradfordcp) [reply azure/modules/iam/main.tf]: Let me know when you add the issue

**Review by bradfordcp — APPROVED** [review]

**Comment by adutra** [comment]
FYI I tried a simple `terraform apply` using this PR and got this error back:

```
│ Error: expanding `default_node_pool`: `node_count`(3) must be equal to or greater than `min_count`(5) when `enable_auto_scaling` is set to `true`
│
│   with module.aks.azurerm_kubernetes_cluster.kubernetes_cluster,
│   on ../modules/aks/main.tf line 16, in resource "azurerm_kubernetes_cluster" "kubernetes_cluster":
│   16: resource "azurerm_kubernetes_cluster" "kubernetes_cluster" {
```

Indeed the plan had:

        + default_node_pool {
            + availability_zones   = [
                + "1",
                + "2",
                + "3",
              ]
            + enable_auto_scaling  = true
            + max_count            = 3
            + max_pods             = (known after apply)
            + min_count            = 5

**Pull request merged**